### PR TITLE
Considering scaling when centering windows

### LINKED ISSUE #14850: WindowStartupLocation="CenterScreen" not working on Linux Mint 21.3 w/ Cinnamon and X11
### Describe the bug

I'm going through the Avalonia.MusicStore tutorial on the website and came to the "Dialog Position and Size" segment where we're instructed to set the WindowStartupLocation in MainWindow.axaml to CenterScreen.

This does not appear to be working as of 11.0.6 or earlier, and as late as 11.0.10 in Linux Mint 21.3 (Virginia) using the Cinnamon Desktop and X11. See the attached image below.

### To Reproduce

On a Linux Mint 21.3 machine, using the Cinnamon Desktop...

1. Create any avalonia.mvvm project. 
2. Go the MainWindow.axaml
3. Add: WindowStartupLocation="CenterScreen" to the window properties
4. Launch/debug the application

### Expected behavior

Expected behavior is that the main window opens in the center of the screen. What actually happens is it opens near the top left corner of the screen. 

### Avalonia version

11.0.10

### OS

Linux

### Additional context

screenshot

### LINKED ISSUE #15905: [Linux] WindowStartupLocation does not respect DesktopScaling factor
### Describe the bug

When `WindowStartupLocation` is set to `CenterScreen` or `CenterOwner`, if the desktop environment has an ambient scaling factor, the window is not centred properly. Its position is set as though its size were unscaled -- so if `DesktopScaling` is 2, for instance, the window is offset to the right and down by a quarter of its size. It gets centred as though the window is half of its actual size, so it extends further to the right and bottom than it should. This has been observed on an Ubuntu 24.04 system using Xorg.

### To Reproduce

Set up a window with `WindowStartupLocation` set to `CenterScreen`. Set the `DesktopScaling` factor to a non-zero value. I actually don't know how this is done -- just that it already has the value `2` on my system, which is a laptop with a 2560x1600 resolution screen. Start the application. Note that the window is not precisely in the centre of the screen.

### Expected behavior

The window should be precisely in the centre of the screen's working area.

### Avalonia version

11.0.10

### OS

Linux

### Additional context

The following work-around seems to work for me:
```
		void CenterWindow()
		{
			if (Screens.Primary is Screen primaryScreen)
			{
				this.Position = new PixelPoint(
					primaryScreen.WorkingArea.X + (int)(primaryScreen.WorkingArea.Width - this.Width * this.DesktopScaling) / 2,
					primaryScreen.WorkingArea.Y + (int)(primaryScreen.WorkingArea.Height - this.Height * this.DesktopScaling) / 2);
			}
		}
```
This can be called from the `Loaded` event of the `Window`.

### PR DESCRIPTION
**Pull request by HermanKirshin**
## What does the pull request do?

Window.Show and ShowDialog calls 
PlatformImpl?.Resize(initialSize, WindowResizeReason.Layout); when scaling on window is not initialized (at least on Linux), so it configures pixel size as if it was 100%.
Then we call SetWindowStartupLocation, and when we do not have owner this returns 1 because we do not have owner and PlatformImpl yet do not have DesktopScaling :
`var scaling = owner?.DesktopScaling ?? PlatformImpl?.DesktopScaling ?? 1;`
And here
`Position = screen.WorkingArea.CenterRect(rect).Position;`
We are attempting to center unscaled window rect within scaled screen bounds, and get incorrect position

## What is the current behavior?

Use Linux
Set desktop scaling to for example 1.5.
Create window with some size and WindowStartupLocation = WindowStartupLocation.CenterScreen
Show window - it is not centered.

## What is the updated/expected behavior with this PR?
Window is shown on the center of the screen

should fix #14850 
should fix #15905

Actually it is reworked version of #12833 that was rolled back in #15304 for release branches, now I hope it should work correctly.

### CODE REVIEW

**Comment by HermanKirshin** [comment]
@cla-avalonia agree

**Comment by HermanKirshin** [comment]
Centering with SizeToContent is still incorrect, working on it

**Comment by HermanKirshin** [comment]
Tested on windows, linux and mac. 


I wasted the whole day for integration tests that were failing on the server but **always** passed successfully on my mac.
(btw I use appium 2, because 1 does not work out of the box, OS version is Sonoma 14.5).

1. WindowOrder_Modal_Dialog_Stays_InFront_Of_Parent_When_Clicking_Resize_Grip
While debugging I noticed that it clicks area in the parent window under the secondary, and not the resize grip. I compared with master and window position and size are the same. So I changed cursor positioning to be relative to the secondary window instead - offset 100 1 is in the middle of the right border. And luckily test did not fail on the server anymore too.

2. WindowOrder_Modal_Dialog_Stays_InFront_Of_Parent, it failed because secondary window index was returned as 2 while it should be 1. Value is taken from textbox and is updated by timer. My suggestion was that it was not updated yet when the test reads the value, so I added a timeout that should wait for at least 1 timer cycle, and looks like it helped.
Maybe setting orderedIndex is somehow delayed for a few ms? But I do not see any problems with window behavior, clicking parent neither changes order of the windows nor steals the focus

4. This one is the hardest.
Horizontal_Changes_Value_Dragging_Thumb_Left and Horizontal_Changes_Value_Dragging_Thumb_Right 

image

However, I got actual values the same 39 and 61 as on the server, but they somehow match expected ones while on server expected are 40 and 60. I added some fractional digits to the textbox that is bound to the slider value and got this.
`Expected: 39.97772828507795, Actual: 38.86`
`Expected: 59.65107873181723, Actual: 61.14`
By the way Actual (slider Value bound to textbox) from run to run is more or less the same, difference is less than 0.1, while Expected (slider value returned by appium) deviates more - difference between 0.5 and 1.0.
I have no clue how this one may be related to the changes in the PR, but I replaced rounding the values with assuming that difference less than 2.0 should be considered as equality. @grokys @maxkatz6 maybe you have some ideas?

File: src/Avalonia.Controls/Window.cs [hunk 1]
```diff
@@ -695,56 +701,120 @@ private void EnsureParentStateBeforeShow(Window owner)
...
+                            x => Closed += x,
+                            x => Closed -= x)
+                        .Take(1)
+                        .Subscribe(_ =>
+                        {
+                            owner!.Activate();
+                            tcs.SetResult((TResult)(_dialogResult ?? default(TResult)!));
+                        });
+                    result = tcs.Task;
+                }
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Window.cs:811]: These changes don't seem to be related to the PR. And it fixes another problem.

**Comment by maxkatz6** [comment]
@emmauss can you check windows side of the PR?

File: src/Avalonia.Controls/Window.cs (hunk 1 above)
Comment (HermanKirshin) [reply src/Avalonia.Controls/Window.cs:811]: Highlighted part does not fix anything - I just merged two almost identical methods (ShowCore and ShowDialog) into one to avoid copypasting all changes twice. There is a separate commit for it - 348222b92270fe47a911f1ef4611e5dbd6c427e6

Comment (maxkatz6) [reply src/Avalonia.Controls/Window.cs:811]: Oh, okay then

**Comment by emmauss** [comment]
> @emmauss can you check windows side of the PR?

Looks fine.

**Review by MrJul — APPROVED** [review]
Already reviewed by @maxkatz6 and @emmauss 
I've re-checked and tested the PR - looks good to me.

Thank you!
